### ISSUE TITLE
[bug] JaxArrays and predictions with non-default X_test parameter

### ISSUE THREAD

**Issue by Kamuish** [dfm/tinygp#85]
Hi @dfm , I think that I found a bug when trying to compute predictions whenever the X_test argument of the GaussianProcess.condition is a non "None" value **and** the X parameter of the GaussianProcess is a Jax array.

If we take some code from the docs (e.g. the quasiseperable tutorial - https://tinygp.readthedocs.io/en/stable/tutorials/quasisep.html) and switch the "t" variable to:

> t = jnp.asarray(t)

And try to run the entire tutorial, we now get an error (pasted as the end of the issue, as it is quite a large traceback). From the API docs, the type-hint of "x" (from the GaussianProcess object) points out that it should be a JaxArray, so this should work, right?

Interestingly, the error does not happen if we condition without specifying the coordinates for prediction:

```
_, cond = build_gp(soln.params).condition(y)

```

- **Traceback**:

```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Final negative log likelihood: 3.79203810708659
Traceback (most recent call last):
  File "/home/amiguel/GP_spectral_modelling/src/falkon_model.py", line 80, in <module>
    _, cond =gp.condition(y, true_t)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/gp.py", line 166, in condition
    alpha, log_prob, mean_value = self._condition(
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/api.py", line 432, in cache_miss
    out_flat = xla.xla_call(
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/core.py", line 1709, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/core.py", line 1721, in call_bind
    outs = top_trace.process_call(primitive, fun, tracers, params)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/core.py", line 614, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/dispatch.py", line 142, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/linear_util.py", line 272, in memoized_fun
    ans = call(fun, *args)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/dispatch.py", line 169, in _xla_callable_uncached
    return lower_xla_callable(fun, device, backend, name, donated_invars,
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/profiler.py", line 206, in wrapper
    return func(*args, **kwargs)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/dispatch.py", line 197, in lower_xla_callable
    jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/profiler.py", line 206, in wrapper
    return func(*args, **kwargs)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1798, in trace_to_jaxpr_final
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(fun, main, in_avals)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1775, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers_)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/gp.py", line 340, in _condition
    mean_value = kernel.matmul(X_test, self.X, alpha)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py", line 145, in matmul
    return self.to_general_qsm(X1, X2) @ y
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py", line 109, in to_general_qsm
    Xs = jax.tree_util.tree_map(lambda x: np.append(x[0], x[:-1]), X2)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/tree_util.py", line 180, in tree_map
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/_src/tree_util.py", line 180, in <genexpr>
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py", line 109, in <lambda>
    Xs = jax.tree_util.tree_map(lambda x: np.append(x[0], x[:-1]), X2)
  File "<__array_function__ internals>", line 5, in append
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4811, in append
    arr = asanyarray(arr)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/jax/core.py", line 470, in __array__
    raise TracerArrayConversionError(self)
jax._src.traceback_util.UnfilteredStackTrace: jax._src.errors.TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
While tracing the function _condition at /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/gp.py:305 for jit, this value became a tracer due to JAX operations on these lines:

  operation a:i64[100] = xla_call[
  call_jaxpr={ lambda ; b:f32[46] c:f64[100]. let
      d:f64[46] = convert_element_type[new_dtype=float64 weak_type=False] b
      e:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 0
      f:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 46
      _:i64[] _:i64[100] g:i64[100] = scan[
        jaxpr={ lambda ; h:f64[46] i:f64[100] j:i64[] k:i64[100] l:i64[100]. let
            m:i64[] = add j 1
            n:i64[100] = add k l
            o:i64[100] = xla_call[
              call_jaxpr={ lambda ; p:i64[100] q:i64[]. let
                  r:i64[100] = div p q
                  s:i64[100] = sign p
                  t:i64[] = sign q
                  u:bool[100] = ne s t
                  v:i64[100] = rem p q
                  w:bool[100] = ne v 0
                  x:bool[100] = and u w
                  y:i64[100] = sub r 1
                  z:i64[100] = xla_call[
                    call_jaxpr={ lambda ; ba:bool[100] bb:i64[100] bc:i64[100]. let
                        bd:i64[100] = select_n ba bc bb
                      in (bd,) }
                    name=vmap(_where)
                  ] x y r
                in (z,) }
              name=vmap(floor_divide)
            ] n 2
            be:bool[100] = lt o 0
            bf:i64[100] = add o 46
            bg:i64[100] = select_n be o bf
            bh:i32[100] = convert_element_type[new_dtype=int32 weak_type=False] bg
            bi:i32[100,1] = broadcast_in_dim[
              broadcast_dimensions=(0,)
              shape=(100, 1)
            ] bh
            bj:f64[100] = gather[
              dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
              fill_value=None
              indices_are_sorted=False
              mode=GatherScatterMode.PROMISE_IN_BOUNDS
              slice_sizes=(1,)
              unique_indices=False
            ] h bi
            bk:i64[100] = bitcast_convert_type[new_dtype=int64] i
            bl:u64[100] = bitcast_convert_type[new_dtype=uint64] i
            bm:bool[100] = eq i 0.0
            bn:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 0
            bo:i64[100] = select_n bm bk bn
            bp:bool[100] = ne i i
            bq:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 9221120237041090560
            br:i64[100] = select_n bp bo bq
            bs:u64[100] = sub 9223372036854775807 bl
            bt:i64[100] = bitcast_convert_type[new_dtype=int64] bs
            bu:bool[100] = lt br 0
            bv:i64[100] = select_n bu br bt
            bw:i64[100] = bitcast_convert_type[new_dtype=int64] bj
            bx:u64[100] = bitcast_convert_type[new_dtype=uint64] bj
            by:bool[100] = eq bj 0.0
            bz:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 0
            ca:i64[100] = select_n by bw bz
            cb:bool[100] = ne bj bj
            cc:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 9221120237041090560
            cd:i64[100] = select_n cb ca cc
            ce:u64[100] = sub 9223372036854775807 bx
            cf:i64[100] = bitcast_convert_type[new_dtype=int64] ce
            cg:bool[100] = lt cd 0
            ch:i64[100] = select_n cg cd cf
            ci:bool[100] = lt bv ch
            cj:i64[100] = xla_call[
              call_jaxpr={ lambda ; ck:bool[100] cl:i64[100] cm:i64[100]. let
                  cn:i64[100] = select_n ck cm cl
                in (cn,) }
              name=vmap(_where)
            ] ci k o
            co:i64[100] = xla_call[
              call_jaxpr={ lambda ; cp:bool[100] cq:i64[100] cr:i64[100]. let
                  cs:i64[100] = select_n cp cr cq
                in (cs,) }
              name=vmap(_where)
            ] ci o l
          in (m, cj, co) }
        length=6
        linear=(False, False, False, False, False)
        num_carry=3
        num_consts=2
        reverse=False
        unroll=1
      ] d c 0 e f
    in (g,) }
  name=searchsorted
] ct cu
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:107 (to_general_qsm)

  operation a:bool[] = lt b b
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:109 (<lambda>)

  operation a:i64[] = add b c
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:109 (<lambda>)

  operation a:i64[] = select_n b c d
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:109 (<lambda>)

  operation a:f32[] = gather[
  dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
  fill_value=None
  indices_are_sorted=True
  mode=GatherScatterMode.PROMISE_IN_BOUNDS
  slice_sizes=(1,)
  unique_indices=True
] b c
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:109 (<lambda>)

(Additional originating lines are not shown.)
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/amiguel/GP_spectral_modelling/src/falkon_model.py", line 80, in <module>
    _, cond =gp.condition(y, true_t)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/gp.py", line 166, in condition
    alpha, log_prob, mean_value = self._condition(
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/gp.py", line 340, in _condition
    mean_value = kernel.matmul(X_test, self.X, alpha)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py", line 145, in matmul
    return self.to_general_qsm(X1, X2) @ y
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py", line 109, in to_general_qsm
    Xs = jax.tree_util.tree_map(lambda x: np.append(x[0], x[:-1]), X2)
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py", line 109, in <lambda>
    Xs = jax.tree_util.tree_map(lambda x: np.append(x[0], x[:-1]), X2)
  File "<__array_function__ internals>", line 5, in append
  File "/home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4811, in append
    arr = asanyarray(arr)
jax._src.errors.TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
While tracing the function _condition at /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/gp.py:305 for jit, this value became a tracer due to JAX operations on these lines:

  operation a:i64[100] = xla_call[
  call_jaxpr={ lambda ; b:f32[46] c:f64[100]. let
      d:f64[46] = convert_element_type[new_dtype=float64 weak_type=False] b
      e:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 0
      f:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 46
      _:i64[] _:i64[100] g:i64[100] = scan[
        jaxpr={ lambda ; h:f64[46] i:f64[100] j:i64[] k:i64[100] l:i64[100]. let
            m:i64[] = add j 1
            n:i64[100] = add k l
            o:i64[100] = xla_call[
              call_jaxpr={ lambda ; p:i64[100] q:i64[]. let
                  r:i64[100] = div p q
                  s:i64[100] = sign p
                  t:i64[] = sign q
                  u:bool[100] = ne s t
                  v:i64[100] = rem p q
                  w:bool[100] = ne v 0
                  x:bool[100] = and u w
                  y:i64[100] = sub r 1
                  z:i64[100] = xla_call[
                    call_jaxpr={ lambda ; ba:bool[100] bb:i64[100] bc:i64[100]. let
                        bd:i64[100] = select_n ba bc bb
                      in (bd,) }
                    name=vmap(_where)
                  ] x y r
                in (z,) }
              name=vmap(floor_divide)
            ] n 2
            be:bool[100] = lt o 0
            bf:i64[100] = add o 46
            bg:i64[100] = select_n be o bf
            bh:i32[100] = convert_element_type[new_dtype=int32 weak_type=False] bg
            bi:i32[100,1] = broadcast_in_dim[
              broadcast_dimensions=(0,)
              shape=(100, 1)
            ] bh
            bj:f64[100] = gather[
              dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
              fill_value=None
              indices_are_sorted=False
              mode=GatherScatterMode.PROMISE_IN_BOUNDS
              slice_sizes=(1,)
              unique_indices=False
            ] h bi
            bk:i64[100] = bitcast_convert_type[new_dtype=int64] i
            bl:u64[100] = bitcast_convert_type[new_dtype=uint64] i
            bm:bool[100] = eq i 0.0
            bn:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 0
            bo:i64[100] = select_n bm bk bn
            bp:bool[100] = ne i i
            bq:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 9221120237041090560
            br:i64[100] = select_n bp bo bq
            bs:u64[100] = sub 9223372036854775807 bl
            bt:i64[100] = bitcast_convert_type[new_dtype=int64] bs
            bu:bool[100] = lt br 0
            bv:i64[100] = select_n bu br bt
            bw:i64[100] = bitcast_convert_type[new_dtype=int64] bj
            bx:u64[100] = bitcast_convert_type[new_dtype=uint64] bj
            by:bool[100] = eq bj 0.0
            bz:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 0
            ca:i64[100] = select_n by bw bz
            cb:bool[100] = ne bj bj
            cc:i64[100] = broadcast_in_dim[broadcast_dimensions=() shape=(100,)] 9221120237041090560
            cd:i64[100] = select_n cb ca cc
            ce:u64[100] = sub 9223372036854775807 bx
            cf:i64[100] = bitcast_convert_type[new_dtype=int64] ce
            cg:bool[100] = lt cd 0
            ch:i64[100] = select_n cg cd cf
            ci:bool[100] = lt bv ch
            cj:i64[100] = xla_call[
              call_jaxpr={ lambda ; ck:bool[100] cl:i64[100] cm:i64[100]. let
                  cn:i64[100] = select_n ck cm cl
                in (cn,) }
              name=vmap(_where)
            ] ci k o
            co:i64[100] = xla_call[
              call_jaxpr={ lambda ; cp:bool[100] cq:i64[100] cr:i64[100]. let
                  cs:i64[100] = select_n cp cr cq
                in (cs,) }
              name=vmap(_where)
            ] ci o l
          in (m, cj, co) }
        length=6
        linear=(False, False, False, False, False)
        num_carry=3
        num_consts=2
        reverse=False
        unroll=1
      ] d c 0 e f
    in (g,) }
  name=searchsorted
] ct cu
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:107 (to_general_qsm)

  operation a:bool[] = lt b b
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:109 (<lambda>)

  operation a:i64[] = add b c
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:109 (<lambda>)

  operation a:i64[] = select_n b c d
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:109 (<lambda>)

  operation a:f32[] = gather[
  dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
  fill_value=None
  indices_are_sorted=True
  mode=GatherScatterMode.PROMISE_IN_BOUNDS
  slice_sizes=(1,)
  unique_indices=True
] b c
    from line /home/amiguel/GP_spectral_modelling/venv/lib/python3.8/site-packages/tinygp/kernels/quasisep.py:109 (<lambda>)

(Additional originating lines are not shown.)
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError

Process finished with exit code 1

```

**Comment by dfm** [comment]
Thanks and good catch! This issue was fixed in https://github.com/dfm/tinygp/pull/81 so the GitHub version of tinygp shouldn't have this problem, but I haven't released a version with this fix yet. I should do that!

**Comment by dfm** [comment]
I released v0.2.2 with this bugfix - let me know if you continue to hit it!

**Comment by Kamuish** [comment]
Things are working fine now, thanks!
